fhir-search/src/main/java/com/ibm/fhir/search/util/ReferenceUtil.java

```diff
@@ -62,18 +62,68 @@ public static ReferenceValue createReferenceValueFrom(Reference ref, String base
         Integer version = null;
 
         if (ref.getReference() != null && ref.getReference().getValue() != null) {
+            return createReferenceValueFrom(ref.getReference().getValue(), ref.getType() != null ? ref.getType().getValue() : null, baseUrl);
+        } else if (ref.getIdentifier() != null && ref.getIdentifier().getValue() != null) {
+            // LOGICAL REFERENCE
+            value = ref.getIdentifier().getValue().getValue();
+            referenceType = ReferenceType.LOGICAL;
+            if (ref.getType() != null) {
+                // We need
+                targetResourceType = ref.getType().getValue();
+            }
+        } else if (ref.getDisplay() != null) {
+            // According to the spec this is still a valid reference. But because it
+            // doesn't refer to anything, we don't need to persist it
+            referenceType = ReferenceType.DISPLAY_ONLY;
+            value = null;
+        } else {
+            // model.Reference currently allows invalid references to be built, so for now we
+            // need to expect and handle this.
+            referenceType = ReferenceType.INVALID;
+            value = null;
+        }
+
+        return new ReferenceValue(targetResourceType, value, referenceType, version);
+    }
+
+    /**
+     * Processes a Reference value from the FHIR model and interprets
+     * it according to https://www.hl7.org/fhir/references.html#2.3.0
+     *
+     * <p>Absolute literal references will be converted to relative references if their base matches baseUrl.
+     *
+     * <p>The resulting ReferenceValue will contain an inferred ReferenceType
+     * and the structure of the ReferenceValue.value will vary accordingly:
+     * <ol>
+     * <li>LITERAL_RELATIVE: the id of the referenced resource</li>
+     * <li>LITERAL_ABSOLUTE: the full URI of the reference</li>
+     * <li>INVALID: null</li>
+     * </ol>
+     *
+     * @param refValue a reference value
+     * @param refType a reference resource type (used for LITERAL_ABSOLUTE only)
+     * @param baseUrl the base URL used to determine whether to convert absolute references to relative references
+     * @return a structured representation of the reference value that varies by its inferred reference type
+     */
+    public static ReferenceValue createReferenceValueFrom(String refValue, String refType, String baseUrl) {
```

Comment:
it not super-clear to me when you would call this variant vs the other createReferenceValueFrom from reading the javadoc

Reply:
Updated the javadoc.